fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/DefaultOCFLObjectSession.java

```diff
@@ -89,85 +115,267 @@ public void write(final String subpath, final InputStream stream) {
+        if (!newInSession(subpath)) {
+            deletePaths.add(subpath);
+        } else if (!hasStagedChanges) {
+            // File is neither in the staged or exists in the head version, so file cannot be found for deletion
+            if (objectDeleted && isStagingEmpty()) {
+                // File can't be found because the object no longer exists
+                throw new PersistentItemNotFoundException(format(
+                        "Could not delete from object %s, it does not exist", objectIdentifier));
+            } else {
+                throw new PersistentItemNotFoundException(format("Could not find %s within object %s",
+                        subpath, objectIdentifier));
+            }
+        }
+    }
+
+    @Override
+    public synchronized void deleteObject() throws PersistentStorageException {
+        assertSessionOpen();
+
+        objectDeleted = true;
+        // Reset state of the object
+        if (!isStagingEmpty()) {
+            cleanupStaging();
+            deletePaths = new HashSet<>();
+        } else if (!ocflRepository.containsObject(objectIdentifier)) {
+            // fail if attempting to delete object that does not exist and has no staged changes
+            throw new PersistentItemNotFoundException(format(
+                    "Cannot delete object %s, it does not exist", objectIdentifier));
+        }
+    }
+
+    private void cleanupStaging() throws PersistentStorageException {
+        try {
+            final var stagingDir = stagingPath.toFile();
+            if (stagingDir.exists()) {
+                FileUtils.cleanDirectory(stagingDir);
+            }
+        } catch (final IOException e) {
+            throw new PersistentStorageException("Unable to cleanup staging path", e);
         }
     }
 
-    /**
-     * {@inheritDoc}
-     */
     private boolean newInSession(final String subpath) {
-        // TODO determine if this subpath exists in the OCFL object
-        return false;
+        // If the object was deleted in this session, then content can only be new
+        if (objectDeleted) {
+            return true;
+        }
+        // If the object isn't created yet, then there is no history for the subpath
+        if (!ocflRepository.containsObject(objectIdentifier)) {
+            return true;
+        }
+        // determine if this subpath exists in the OCFL object
+        return !ocflRepository.getObject(ObjectVersionId.head(objectIdentifier))
+                .containsFile(subpath);
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
-    public InputStream read(final String subpath) {
-        if (hasStagedChanges(subpath)) {
-            // TODO read the staged version of the file
-        } else {
-            // TODO read the head version of the file from the ocfl object
+    public InputStream read(final String subpath) throws PersistentItemNotFoundException {
```

Comment:
Do the `read` methods need session checks?

Reply:
I was thinking it wouldn't do any harm to read from a closed session since it'd fall back to the head version, but its probably best to force consistency by disallowing session reuse. I'll change it